### PR TITLE
Align jqLite's attr method with jQuery

### PR DESCRIPTION
**Pull request by mgol**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Refactors & breaking API changes.

**What is the current behavior? (You can also link to an open issue here)**
Boolean attribute getters take properties into account and setters modify them. All falsy values for boolean attribute setters remove the attribute and set the property to false. The `null` value in a setter (for all attributes, not only boolean ones) sets the attribute value to the string `"null"` instead of removing it.

**What is the new behavior (if this is a feature change)?**
Boolean attribute getters/setters don't touch properties. The `null` value in a setter (for all attributes, not only boolean ones) removes the attribute. The `false` value removes it only for boolean attributes. Other defined values for boolean attributes set the value to the lowercased attribute name.

**Does this PR introduce a breaking change?**
Yes.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~ - only in a form of commit messages with their `BREAKING CHANGES` sections; is anything else needed?

**Other information**:

### CODE REVIEW

**Comment by Narretz** [comment]
There are a few failures when running the tests with jquery 2.2, which is probably expected. How do we handle this?

**Comment by mgol** [comment]
@Narretz I fixed the test: I'm just skipping that one in jQuery 2.x as this was a jQuery 3.0 change.

BTW, most other changes here are aligning even with older jQuery versions, this is not just about 3.0 changes.

**Comment by mgol** [comment]
Tests pass, PR ready for review.

File: test/jqLiteSpec.js [hunk 1]
```diff
@@ -635,6 +635,56 @@ describe('jqLite', function() {
       expect(select.attr('multiple')).toBe('multiple');
     });
 
+    it('should set respective properties when getting/setting boolean attributes', function() {
```
Comment (gkalpak) [inline test/jqLiteSpec.js]: As far as I understand, it is actually the browser that does that, not jqLite/`.attr()`. 
We should either remove the test (if we don't consider it to be jqLite's job, or make sure jqLite always sets the properties (instead of relying on the browsers to do it).

Comment (gkalpak) [inline test/jqLiteSpec.js]: Also, `getting` seems not related here.

File: test/jqLiteSpec.js [hunk 2]
```diff
@@ -635,6 +635,56 @@ describe('jqLite', function() {
...
+
+      div[0].multiple = true;
+      expect(div.attr('multiple')).toBe(undefined);
+
+      div.attr('multiple', 'multiple');
+      div[0].multiple = false;
+      expect(div.attr('multiple')).toBe('multiple');
+    });
+
+    it('should not set properties when settting respective boolean attributes', function() {
```
Comment (gkalpak) [inline test/jqLiteSpec.js]: Too much `t` in `setttings` :smiley:

File: src/jqLite.js [hunk 3]
```diff
@@ -638,33 +638,32 @@ forEach({
   },
 
   attr: function(element, name, value) {
+    var ret;
```
Comment (gkalpak) [inline src/jqLite.js]: <sub>Nit: I prefer declaring vars in the block in which they are used.</sub>

File: src/jqLite.js [hunk 4]
```diff
@@ -638,33 +638,32 @@ forEach({
...
-          element[name] = false;
-          element.removeAttribute(lowercasedName);
-        }
+    var isBooleanAttr = BOOLEAN_ATTR[lowercasedName];
+
+    if (isDefined(value)) {
+      // setter
+
+      if (value === null || (value === false && isBooleanAttr)) {
+        element.removeAttribute(name);
```
Comment (gkalpak) [inline src/jqLite.js]: Previously, for boolean attributes, we used the `lowercasedName`. Does it make a difference?

File: src/jqLite.js [hunk 5]
```diff
@@ -638,33 +638,32 @@ forEach({
...
-        return (element[name] ||
-                 (element.attributes.getNamedItem(name) || noop).specified)
-               ? lowercasedName
-               : undefined;
+        element.setAttribute(name, isBooleanAttr ? lowercasedName : value);
+      }
+    } else {
+      // getter
+
+      ret = element.getAttribute(name);
```
Comment (gkalpak) [inline src/jqLite.js]: Is there any chance `getAttribute(name)` could return `undefined` (on any browser)?

File: src/jqLite.js [hunk 6]
```diff
@@ -638,33 +638,32 @@ forEach({
...
+
+      if (isBooleanAttr && ret !== null) {
+        ret = lowercasedName;
       }
-    } else if (isDefined(value)) {
-      element.setAttribute(name, value);
-    } else if (element.getAttribute) {
```
Comment (gkalpak) [inline src/jqLite.js]: Previously, there was this `element.getAttribute` check, which is now removed.
Was it necessary for IE<9 browsers only? It would be useful to have a comment (e.g. in the PR) explaining why it is not necessary any more.

File: src/jqLite.js [hunk 7]
```diff
@@ -638,33 +638,32 @@ forEach({
...
-      // the extra argument "2" is to get the right thing for a.href in IE, see jQuery code
-      // some elements (e.g. Document) don't have get attribute, so return undefined
-      var ret = element.getAttribute(name, 2);
```
Comment (gkalpak) [inline src/jqLite.js]: Same as above: It would be useful to have a comment (e.g. in the PR) explaining why this is not necessary any more (assuming it isn't 😛).

**Review by gkalpak — COMMENTED** [review]
I have left a couple of (mostly minor) comments. 

One thing that I believe is not pointed out explicitly enough in the BC notices is the fact that boolean attribute getters do **not** rely on the value of the corresponding property (as was previously the case).

From the commit message:

> Previously, all boolean attributes were reflected into properties in a setter and from a property in a getter, even on elements that don't treat those attributes in a special way. Now it's only happening for the elements where browsers do the conversion automatically.

This sounds to me like boolean attribute getters will still take the property into account, but only on browsers that do the conversion automatically (which is not the case).

File: test/jqLiteSpec.js (hunk 1 above)
Comment (mgol) [inline test/jqLiteSpec.js]: IMO it often does make sense to test things like that; in a way that we should have had a test like that in `1.5.x` and it should remain passing after we remove our own reflection. In this way we'd know if a buggy browser broke it and we need to patch it ourselves.

But maybe here we wouldn't do the patching either way so I'd be fine either way.

File: src/jqLite.js (hunk 4 above)
Comment (mgol) [inline src/jqLite.js]: It shouldn't. Also, jQuery removes `name`, not `lowercasedName`.

File: src/jqLite.js (hunk 5 above)
Comment (mgol) [inline src/jqLite.js]: Good catch. It's needed for `window` doesn't have `getAttribute`. In such cases [jQuery falls back from `attr` to `prop`](https://github.com/jquery/jquery/blob/3.1.1/src/attributes/attr.js#L37-L39); should we do the same here?

File: src/jqLite.js (hunk 6 above)
Comment (mgol) [inline src/jqLite.js]: Good catch. It's needed for `window` doesn't have `getAttribute`. In such cases [jQuery falls back from `attr` to `prop`](https://github.com/jquery/jquery/blob/3.1.1/src/attributes/attr.js#L37-L39); should we do the same here?

File: src/jqLite.js (hunk 5 above)
Comment (mgol) [inline src/jqLite.js]: I don't think so but if that happened then what do you propose? jQuery only checks for `null` to convert it to `undefined`, if a browser returned `undefined` we'd already have what we want.

File: test/jqLiteSpec.js (hunk 1 above)
Comment (gkalpak) [inline test/jqLiteSpec.js]: I was under the impression that we wouldn't do the patching even if a browser/version failed to do the right thing. If the plan is to do the patching in that case, then I would rather be on the afe side and set the properties anyway. (But if jQuery doesn't do it, I think it is better for jqLite not to do it either.)

I don't feel strongly about keeping the test around, but a different description that makes it clear we expect _the browser_ to do it would probably make things a little easier fr our future selves, if something breaks :smiley:

File: src/jqLite.js (hunk 6 above)
Comment (gkalpak) [inline src/jqLite.js]: SGTM

File: src/jqLite.js (hunk 5 above)
Comment (gkalpak) [inline src/jqLite.js]: [This line][1] was using a `!= null` check, so I assumed `undefined` is a possible return value. If that is the case, then we shouldn't set `ret = lowercasedName` when `ret === undefined`, but afaict `getAttribute()` shouldn't ever return `undefined` (in which case we are fine).

[1]: https://github.com/angular/angular.js/pull/15181/commits/6dc5e761bdee39baa9718316acc0c764f5130d25#diff-8e0d9e5c18fe4f1050484080525d4367R654

File: test/jqLiteSpec.js (hunk 1 above)
Comment (mgol) [inline test/jqLiteSpec.js]: OK, I removed the test.

File: src/jqLite.js (hunk 6 above)
Comment (mgol) [inline src/jqLite.js]: Done.

Comment (mgol) [inline src/jqLite.js]: It was still failing in a setter which jQuery doesn't so I fixed that as well. I decided to not fall back to prop for now, though, but just return as this feature is not very well tested & documented in jQuery.

File: src/jqLite.js (hunk 7 above)
Comment (mgol) [inline src/jqLite.js]: Done.

File: src/jqLite.js [hunk 8]
```diff
@@ -56,7 +56,7 @@
  * - [`after()`](http://api.jquery.com/after/)
  * - [`append()`](http://api.jquery.com/append/)
  * - [`attr()`](http://api.jquery.com/attr/) - Does not support functions as parameters
- * - [`bind()`](http://api.jquery.com/bind/) - Does not support namespaces, selectors or eventData
+ * - (_deprecated_, use `on()`) [`bind()`](http://api.jquery.com/bind/) - Does not support namespaces, selectors or eventData
```
Comment (mgol) [inline src/jqLite.js]: Is that a good way to mark methods as deprecated? Thoughts?

Comment (mgol) [inline src/jqLite.js]: That's how it looks like:
screen shot 2016-10-05 at 12 42 29

Comment (mgol) [inline src/jqLite.js]: Changed.

Comment (gkalpak) [inline src/jqLite.js]: We should alse:

1. Document deprecations with the release in which they will be removed (usually the next minor release).
2. Place an issue in the appropriate milestone to remind us to do so.

File: test/jqLiteSpec.js [hunk 9]
```diff
@@ -2256,4 +2334,19 @@ describe('jqLite', function() {
       expect(onLoadCallback).toHaveBeenCalledOnce();
     });
   });
+
+
+  describe('bind()/unbind()', function() {
```
Comment (gkalpak) [inline test/jqLiteSpec.js]: <sub>Super-nit: Other descriptions don't have parenthesis :grin:</sub>

File: src/jqLite.js (hunk 3 above)
Comment (petebacondarwin) [inline src/jqLite.js]: I prefer small functions so that there is not much difference between the two approaches :-P

File: test/jqLiteSpec.js (hunk 9 above)
Comment (petebacondarwin) [inline test/jqLiteSpec.js]: Don't worry about it

**Comment by petebacondarwin** [comment]
LGTM

File: test/jqLiteSpec.js (hunk 9 above)
Comment (mgol) [inline test/jqLiteSpec.js]: Already changed locally! :D

**Review by gkalpak — APPROVED** [review]
LGTM
<sub>(Don't forget the issue for removing `bind/unbind' :smiley:)</sub>

**Comment by gkalpak** [comment]
(Also, feel free to tweak the commit message as per https://github.com/angular/angular.js/pull/15181#pullrequestreview-2284088 :smiley:.)

**Comment by petebacondarwin** [comment]
@mgol - feel free to tweak as suggested by @gkalpak  and then merge. Thanks for your work on this.

**Comment by mgol** [comment]
Removal issue: #15221.
